go/timedlock/rw_lock.go

```diff
@@ -0,0 +1,191 @@
+package timedlock
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+)
+
+var timerPool = &sync.Pool{}
+
+const MAX_RLOCK = 1000
+
+var TIMEOUT_ERROR = errors.New("Timeout occured")
+
+type RWLock struct {
+	internalLock *sync.Mutex
+
+	lockChan  chan int
+	rlockChan chan int
+
+	// A counter to track the amount of allocated locks for providing the next ID
+	c uint64
+
+	// The timeout for Lock operations
+	Timeout time.Duration
+
+	// The timeout for RLock operations
+	RTimeout time.Duration
+
+	// Whether or not this should panic on timeouts
+	// Otherwise, it will just release the lock
+	Panic bool
+
+	// Whether or not the errors should be printed
+	PrintErrors bool
+}
+
+func NewRWLock() *RWLock {
+	l := &RWLock{
+		lockChan:     make(chan int, 1),
+		rlockChan:    make(chan int, MAX_RLOCK),
+		Timeout:      10 * time.Second,
+		RTimeout:     10 * time.Second,
+		internalLock: &sync.Mutex{},
+		Panic:        true,
+		PrintErrors:  false,
+		c:            1,
+	}
+	l.lockChan <- 1
+	for i := 0; i < MAX_RLOCK; i++ {
+		l.rlockChan <- 1
+	}
+	return l
+}
+
+func (l *RWLock) getTimer(d time.Duration) *time.Timer {
+	o := timerPool.Get()
+	if o == nil {
+		return time.NewTimer(d)
+	} else {
+		t := o.(*time.Timer)
+		t.Reset(d)
+		return t
+	}
+}
+
+func (l *RWLock) releaseTimer(t *time.Timer) {
+	t.Stop()
+	timerPool.Put(t)
+}
+
+func (l *RWLock) getNextId() uint64 {
+	l.internalLock.Lock()
+	defer l.internalLock.Unlock()
+	l.c++
+	// 0 isn't an acceptable ID
+	if l.c == 0 {
+		l.c++
+	}
+	return l.c
+}
+
+func (l *RWLock) Lock() (uint64, error) {
+	id := l.getNextId()
+	stopAt := time.Now().Add(l.Timeout)
+	timeoutTimer := l.getTimer(l.Timeout)
+	defer l.releaseTimer(timeoutTimer)
+
+	select {
+	case <-timeoutTimer.C:
+		l.handleTimeout(id, "Timeout obtaining or releasing lock that came from: \n")
+		return 0, TIMEOUT_ERROR
+	case <-l.lockChan:
+		timeoutTimer.Stop()
+		for time.Now().Before(stopAt) {
+			l.internalLock.Lock()
+			if len(l.rlockChan) == MAX_RLOCK {
+				l.internalLock.Unlock()
+				return id, nil
+			} else {
+				l.internalLock.Unlock()
+			}
```

Comment:
In this busy loop call time.Sleep() or runtime.Goscheld() after a few iterations do a avoid pegging the CPU to 100%